src/components/RepeatableReferenceField.tsx

```diff
@@ -0,0 +1,157 @@
+// @ts-nocheck
+import { React, useState, useEffect } from 'react';
+import {
+  Button,
+  EditorToolbarButton,
+  SkeletonBodyText,
+  SkeletonContainer,
+  Table,
+  TableBody,
+  TableRow,
+  TableCell,
+  TextInput
+} from '@contentful/forma-36-react-components';
+import { FieldExtensionSDK } from 'contentful-ui-extensions-sdk';
+
+interface FieldProps {
+  sdk: FieldExtensionSDK;
+}
+
+const RepeatableReferenceField = (props: FieldProps) => {
+  const fieldValue = props.sdk.field.getValue();
+  const [rows, setRows] = useState(fieldValue || []);
+  const instanceParameters = props.sdk.parameters.instance;
+  const referenceKey = instanceParameters.referenceKey || 'id';
+  const textKey = instanceParameters.textKey || 'text';
+  const textLabel = instanceParameters.textLabel || 'Text';
+  const contentTypes = instanceParameters.contentTypes
+    ? instanceParameters.contentTypes.split(/\s*,\s*/g)
+    : null;
+
+  // use contentful's builtin auto-resizer 
+  useEffect(() => {
+    props.sdk.window.startAutoResizer();
+  })
+
+  // check for unresolved names and fetch them from contenful if neccessary
+  useEffect(() => {
+    const unpopulatedRows = rows.filter((row) => !row.name);
+    if (!unpopulatedRows.length) {
+      return;
+    }
+
+    const referencedIds = rows.map((row) => row[referenceKey]);
+    props.sdk.space.getEntries({ 'sys.id[in]': referencedIds }).then((queryResult) => {
+      let populatedRows = rows.map((row) => {
+        const resultForCurrentRow = queryResult.items.filter((entry) => entry.sys.id === row[referenceKey]).pop();
+        return {
+          name: resultForCurrentRow ? resultForCurrentRow.fields.title['en-US'] : '',
+          ...row
+        }
+      });
+      setRows(populatedRows);
+    });
+  }, [rows, props.sdk.space, referenceKey]);
+
+  // update contentful field value whenever rows data changes
+  useEffect(() => {
+    const sanitizedRows = rows.map((row) => {
+      const sanitizedRow = {};
+      sanitizedRow[textKey] = row[textKey];
+      sanitizedRow[referenceKey] = row[referenceKey];
+      return sanitizedRow;
+    });
+    props.sdk.field.setValue(sanitizedRows);
+  }, [rows, props.sdk.field, referenceKey, textKey]);
+
+  // open entry selection dialog and append selected entries to the end of our list
+  const onAddButtonClicked = () => {
+    const options = {};
+    if (contentTypes) {
+      options.contentTypes = contentTypes;
+    }
+    props.sdk.dialogs.selectMultipleEntries(options)
+      .then((selectedRows) => {
+        setRows([
+          ...rows,
+          ...selectedRows.map((row) => {
+            const rowData = {
+              key: `${row.sys.id}-${Math.floor(Math.random() * 100000)}`
```

Comment:
Is the reason you're creating this random ID because you can have multiple list items which are the same reference? If that's not that case, I would simply go with using the `row.sys.id` so the `key` is consistent for React to handle.

Reply:
Yeah, exactly. It made sense for the ingredients because ingredients sometimes occurred more than once in a recipe and it was annoying to go back to the first reference and add an amount... I think it can still make sense for the future. Imagine e.g. a list of orders in an online shop where two customers order the same item. Do you have an idea for a more consistent key schema?

Reply:
In my repo, I'm doing something similar and using this:
https://www.npmjs.com/package/uuid

I think it will create more consistent uuids with less potential for clashing.

Reply:
Thanks for the suggestion! I'll try that then and let you know once it's ready